File: 백준/Gold/2632. 피자판매/피자판매.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;
import java.util.StringTokenizer;

public class Main {

    static int target;
    static int m, n;
    static int[] pizzaA, pizzaB;
    static Map<Integer, Integer> pizzaAMap, pizzaBMap;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        target = Integer.parseInt(br.readLine());
        st = new StringTokenizer(br.readLine());
        m = Integer.parseInt(st.nextToken());
        n = Integer.parseInt(st.nextToken());

        pizzaA = new int[m];
        pizzaB = new int[n];

        for (int i = 0; i < m; i++) {
            pizzaA[i] = Integer.parseInt(br.readLine());
        }

        for (int i = 0; i < n; i++) {
            pizzaB[i] = Integer.parseInt(br.readLine());
        }

        pizzaAMap = new HashMap<>();
        pizzaBMap = new HashMap<>();

        createMap(m, pizzaA, pizzaAMap);
        createMap(n, pizzaB, pizzaBMap);

        int answer = 0;

        if (pizzaAMap.containsKey(target)) {
            answer += pizzaAMap.get(target);
        }

        if (pizzaBMap.containsKey(target)) {
            answer += pizzaBMap.get(target);
        }

        for (Integer pizzaASize : pizzaAMap.keySet()) {
            int pizzaBSize = target - pizzaASize;
            if (!pizzaBMap.containsKey(pizzaBSize)) {
                continue;
            }

            answer += pizzaAMap.get(pizzaASize) * pizzaBMap.get(pizzaBSize);

        }

        System.out.println(answer);
    }

    private static void createMap(int length, int[] pizza, Map<Integer, Integer> pizzaMap) {
        for (int start = 0; start < length; start++) {
            int tempSum = 0;
            for (int size = 0; size < length; size++) {
                int index = (start + size) % length;
                tempSum += pizza[index];
                if (pizzaMap.containsKey(tempSum)) {
                    pizzaMap.put(tempSum, pizzaMap.get(tempSum) + 1);
                } else {
                    pizzaMap.put(tempSum, 1);
                }
            }
        }

        pizzaMap.put(Arrays.stream(pizza).sum(), 1);
    }
}

File: 백준/Gold/2632. 피자판매/피자판매.py
import sys
from collections import defaultdict

input = sys.stdin.readline


targetSize = int(input())
m, n = map(int, input().split())
pizzaA = [int(input()) for _ in range(m)]
pizzaB = [int(input()) for _ in range(n)]

pizzaASizes = defaultdict(int)
pizzaBSizes = defaultdict(int)

for start in range(m):
    size = 0
    for length in range(m - 1):
        idx = (start + length) % m
        size += pizzaA[idx]
        pizzaASizes[size] += 1

pizzaASizes[sum(pizzaA)] += 1

for start in range(n):
    size = 0
    for length in range(n - 1):
        idx = (start + length) % n
        size += pizzaB[idx]
        pizzaBSizes[size] += 1

pizzaBSizes[sum(pizzaB)] += 1

answer = pizzaASizes[targetSize] + pizzaBSizes[targetSize]

for pizzaASize in pizzaASizes:
    pizzaBSize = targetSize - pizzaASize
    answer += pizzaASizes[pizzaASize] * pizzaBSizes[pizzaBSize]

print(answer)